### PR TITLE
Model application name and system tag checks

### LINKED ISSUE #422: Model checks for `ApplicationName` and `SystemTag` in the specs
The `ApplicationName` is part of the `SoftwareVersion` and should be an ascii string of length 12 or less. The `SystemTag` is used in the update metadata to tag metadata for each target system, it should be an ascii string of length 10 or less. Both of these could be modelled in the specs using abstract functions like `appNameValid :: AppName -> Bool`.

### PR DESCRIPTION
**Pull request by dnadales**
Closes #422.

### CODE REVIEW

File: byron/ledger/formal-spec/update-mechanism.tex [hunk 1]
```diff
@@ -130,6 +132,7 @@ \subsection{Update proposals}
       \fun{upSwVer} & \UProp \to \SWVer & \text{software-version update proposal}\\
       \fun{upSig} & \UProp \to \Sig & \text{update proposal signature}\\
       \fun{upSigdata} & \UProp \to \UPropSD & \text{update proposal signed data}\\
+      \fun{upSTags} & \powerset{\STag} & \text{update proposal system tags}
```
Comment (nc6) [inline byron/ledger/formal-spec/update-mechanism.tex]: Might be nice to have some more definition of what this is in prose.

**Comment by dnadales** [comment]
> There's no Haskell implementation of these two things. They shouldn't be hard to add.

This PR concerns the formal spec only.

**Comment by mdimjasevic** [comment]
> 
> This PR concerns the formal spec only.

To avoid further drifting apart of the LaTeX and Haskell counterparts of the spec, I'd suggest doing this in a lockstep fashion.

File: byron/ledger/formal-spec/update-mechanism.tex (hunk 1 above)
Comment (dnadales) [reply byron/ledger/formal-spec/update-mechanism.tex]: Good point. Done.

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 2]
```diff
@@ -244,8 +257,11 @@ instance STS UPPVV where
         pvCanFollow nv pv ?! CannotFollowPv
         canUpdate pps up ?! CannotUpdatePv
         nv `notElem` (fst <$> Map.elems rpus) ?! AlreadyProposedPv
+        all sTagValid (up ^. upSTags) ?! InvalidSystemTags
         return $! rpus ⨃ Map.singleton pid (nv, ppsn)
     ]
+    where
+      sTagValid tag = all isAscii tag && length tag <= 10
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: I'm not too familiar with the semantics of system tags, but shouldn't `STag` be a small finite set? I suppose that "FunnySys" shouldn't pass this check validity, yet there's nothing in the current implementation preventing it from passing the check.

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 3]
```diff
@@ -206,17 +213,22 @@ instance STS UPSVV where
...
     [ do
         TRC (avs, raus, up) <- judgmentContext
         let SwVer an av = up ^. upSwVer
+        apNameValid an ?! InvalidApplicationName
         svCanFollow avs (an,av) ?! CannotFollowSv
         (an, av) `notElem` Map.elems raus ?! AlreadyProposedSv
         return $! raus ⨃ [(up ^. upId, (an, av))]
     ]
+    where
+      apNameValid (ApName n) = all isAscii n && length n <= 12
```
[the remark this answers is not in the document]
Comment (dnadales) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: Good catch. It should be `<=` I adapted the prose.

File: byron/ledger/executable-spec/src/Ledger/Update.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dnadales) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: > I'm not too familiar with the semantics of system tags, but shouldn't STag be a small finite set? 

Not at the moment and modeling this is outside of the scope of what this PR is addressing.

**Pull request merged**